File: app/services/study_service.py
from app.domain.models import Study, new_id

class StudyService:
    def __init__(self, measurement_registry):
        self.registry = measurement_registry
        self.studies = {}

    def create(self, title, measurement_ids=None, seed=20260922):
        ids = measurement_ids or list(self.registry.keys())[:5]
        unknown = [x for x in ids if x not in self.registry]
        if unknown:
            raise ValueError(f"Unknown measurements: {unknown}")
        s = Study(new_id("study"), title, measurement_ids=ids, seed=seed)
        self.studies[s.study_id] = s
        return s

    def freeze(self, study_id):
        s = self.studies[study_id]
        s.status = "FROZEN"
        return s

File: src/server/studyService.ts
import { Study } from "../types.js";
import { REGISTRY } from "./registry.js";
import { randomUUID } from "crypto";

export class StudyService {
  private studies: Map<string, Study> = new Map();

  create(title?: string, measurementIds?: string[], seed = 20260922): Study {
    const ids = measurementIds || Object.keys(REGISTRY).slice(0, 5);
    const unknown = ids.filter((id) => !REGISTRY[id]);
    if (unknown.length > 0) {
      throw new Error(`Unknown measurements: ${unknown.join(", ")}`);
    }

    const studyId = `study_${randomUUID().replace(/-/g, "").slice(0, 12)}`;
    const study: Study = {
      study_id: studyId,
      title: title || "DNEM Demo Study",
      version: "v7.7",
      status: "DRAFT",
      measurement_ids: ids,
      seed,
      created_at: Date.now() / 1000,
    };
    this.studies.set(studyId, study);
    return study;
  }

  freeze(studyId: string): Study {
    const study = this.studies.get(studyId);
    if (!study) {
      throw new Error(`Study not found: ${studyId}`);
    }
    study.status = "FROZEN";
    return study;
  }

  get(studyId: string): Study | undefined {
    return this.studies.get(studyId);
  }

  list(): Study[] {
    return Array.from(this.studies.values());
  }
}

export const studyService = new StudyService();
